src/activity_handlers.cpp

```diff
@@ -864,66 +865,93 @@ void activity_handlers::pulp_do_turn( player_activity *act, player *p )
 {
     const tripoint &pos = act->placement;
 
-    // numbers logic: a str 8 character with a butcher knife (4 bash, 18 cut)
-    // should have at least a 50% chance of damaging an intact zombie corpse (75 volume).
-    // a str 8 character with a baseball bat (28 bash, 0 cut) should have around a 25% chance.
-
     int cut_power = p->weapon.type->melee_cut;
-    // stabbing weapons are a lot less effective at pulping
+    // Stabbing weapons are a lot less effective at pulping
     if( p->weapon.has_flag("STAB") || p->weapon.has_flag("SPEAR") ) {
         cut_power /= 2;
     }
+
+    // Slicing weapons are a moderately less effective at pulping
+    if( p->weapon.has_flag("SLICE") ) {
+        cut_power = cut_power * 2 / 3;
+    }
     ///\EFFECT_STR increases pulping power, with diminishing returns
-    double pulp_power = sqrt((double)(p->str_cur + p->weapon.type->melee_dam)) *
-        std::min(1.0, sqrt((double)(cut_power + 1)));
-    ///\EFFECT_STR caps pulping power
-    pulp_power = std::min(pulp_power, (double)p->str_cur);
-    pulp_power *= 20; // constant multiplier to get the chance right
+    float pulp_power = sqrt( (p->str_cur + p->weapon.type->melee_dam) * ( cut_power + 1.0f ) );
+    // Multiplier to get the chance right + some bonus for survival skill
+    pulp_power *= 40 + p->get_skill_level( skill_survival ) * 5;
+
+    const int mess_radius = p->weapon.has_flag("MESSY") ? 2 : 1;
 
     int moves = 0;
     int &num_corpses = act->index; // use this to collect how many corpse are pulped
     auto corpse_pile = g->m.i_at(pos);
     for( auto corpse = corpse_pile.begin(); corpse != corpse_pile.end(); ++corpse ) {
+        if( !corpse->is_corpse() || !corpse->get_mtype()->has_flag( MF_REVIVES )  ) {
```

Comment:
This check isn't repeated when registering the activity, so if you smash a non-reviving corpse, it will tell you, "The corpse moved before you could finish smashing it!" instead of "nothing there to smash.".
I'll probably fix this with my in-progress merge, just noting it as a FYI and in case I get pulled away from my merge.